### DISCUSSION TITLE
Any well-known way to deploy sw360 using an official image

### DISCUSSION THREAD

**Discussion by davidecavestro** [eclipse-sw360/sw360#2177] [Q&A]
I've seen the [recommended setup](https://github.com/eclipse-sw360/sw360/blob/main/README_DOCKER.md) is based on self-building the docker image.

The [official compose](https://github.com/eclipse-sw360/sw360/blob/main/docker-compose.yml) uses the image _ghcr.io/eclipse-sw360/sw360:latest_ which is not published, but the locally built image has that name.
OTOH there's an image hosted at _ghcr.io/eclipse/sw360:latest_ (please note the name differs).

At the moment locally building the image fails both on main and on the last tag (I can provide details if interested, but doesn't really matter here).

When I adapt the compose to run with the image at _ghcr.io/eclipse/sw360:latest_ I get some errors at spring startup (i.e. _IdleConnectionMonitor$CleanupTask_ rejected, with _pool size = 0_ or _InvokerFilter cannot be found by ... liferay-theme_ and so on), so before going too much into depth, I'd like to know:

**is there any well known way to deploy sw360 using an official image with some predefined config?**

**Comment by heliocastro** [comment]
The name of published image comes from previous repository name change, and is a bug. Thanks for catching up.
The compose image is correct, so fixing upstream will be fixing the compose.
I will update this as soon a 17.0.x release is done.

**Comment by heliocastro** [comment]
Main package is public now:
https://github.com/eclipse-sw360/sw360/pkgs/container/sw360

I will raise the decision over latest on next sw360 meeting.
Later tonight, schedule build will happens, and you will at least have nightly tagged.

**Reply by davidecavestro** [reply]
ok, so I tried ghcr.io/eclipse-sw360/sw360:17.0.0-SNAPSHOT and Tomcat complains for

> sw360               |   Caused by: java.lang.NoSuchFieldException: modifiers
> sw360               |           at java.base/java.lang.Class.getDeclaredField(Class.java:2610)
> sw360               |           at com.liferay.petra.reflect.ReflectionUtil.<clinit>(ReflectionUtil.java:157)
> sw360               |           ... 47 more
> 

I mention this here just because at a first sight it seems like a jdk compatibility issue of the packaged liferay.

**Reply by heliocastro** [reply]
Liferay filters... This not affects execution, but is really an annoyance to us since some time.

**Reply by davidecavestro** [reply]
Probably I missed something, cause I always see the root application as not started.

I consistently reproduce it at https://labs.play-with-docker.com
simply issuing the following commands
```
# get the compose
curl -sSL -o compose.yml https://raw.githubusercontent.com/eclipse-sw360/sw360/main/docker-compose.yml
# change image version
sed -i s/sw360:latest/sw360:17.0.0-SNAPSHOT/g compose.yml
# start it
docker compose -f compose.yml up
```
![image](https://github.com/eclipse-sw360/sw360/assets/447917/3de0a376-eb00-41c5-bec8-0f54bfca12b0)

**Comment by heliocastro** [comment]
Can you try in a local system instead of play-with-docker
I tried same and can easily see problems with database connections.

**Reply by davidecavestro** [reply]
> Can you try in a local system instead of play-with-docker
> I tried same and can easily see problems with database connections.
> 

Actually I tried on my local Linux box before, as play-with-docker was just an attempt to share some steps to repro.

Opening a shell into the local sw360 container I listed the contents of the _shared_ subfolder within tomcat and saw duplicated jars having different versions.
I'd say from _catalina.properties_ entries that the _shared_ classloader is enabled, so could that lead to clashes? (sorry I don't have the pc at hand now to copy-paste exact details)

**Comment by heliocastro** [comment]
Ok, this is interesting, duplicated libraries ( version wise ) was a problem in the past and that's why we went to the shared approach. 
I'm rebuilding an entire thing from scratch on mu local machine to check it out.

**Reply by davidecavestro** [reply]
just to give an idea about first 20 jars within the shared folder
```
docker run --rm -it --entrypoint /bin/bash \
  ghcr.io/eclipse-sw360/sw360:17.0.0-SNAPSHOT -c \
  'find /app/sw360/tomcat/shared/ -name "*.jar" | sort | head -n 20'
```
![image](https://github.com/eclipse-sw360/sw360/assets/447917/f1c85904-c131-4fd2-89a3-21e8db10fb9c)

**Reply by davidecavestro** [reply]
Please also note that the exception having root cause `java.lang.NoSuchFieldException: modifiers`
is marked as _SEVERE_ and seems to prevent the ROOT app to boot:
```
sw360  | 20-Nov-2023 08:11:09.659 SEVERE [main] org.apache.catalina.startup.HostConfig.deployDescriptor Error deploying deployment descriptor [/app/sw360/tomcat-9.0.56/conf/Catalina/localhost/ROOT.xml]
sw360  |        java.lang.IllegalStateException: Error starting child
sw360  |                at org.apache.catalina.core.ContainerBase.addChildInternal(ContainerBase.java:729)
...
sw360  |                at org.apache.catalina.startup.Bootstrap.main(Bootstrap.java:476)
sw360  |        Caused by: org.apache.catalina.LifecycleException: Failed to start component [StandardEngine[Catalina].StandardHost[localhost].StandardContext[]]
sw360  |                at org.apache.catalina.util.LifecycleBase.handleSubClassException(LifecycleBase.java:440)
sw360  |                at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:198)
sw360  |                at org.apache.catalina.core.ContainerBase.addChildInternal(ContainerBase.java:726)
sw360  |                ... 37 more
sw360  |        Caused by: java.lang.ExceptionInInitializerError
sw360  |                at com.liferay.portal.asm.ASMWrapperUtil.<clinit>(ASMWrapperUtil.java:272)
...
sw360  |                at com.liferay.shielded.container.internal.proxy.ProxyFactory.createASMWrapper(ProxyFactory.java:51)
sw360  |                at com.liferay.shielded.container.internal.ShieldedContainerServletContainerInitializer.onStartup(ShieldedContainerServletContainerInitializer.java:59)
sw360  |                at org.apache.catalina.core.StandardContext.startInternal(StandardContext.java:5219)
sw360  |                at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:183)
sw360  |                ... 38 more
sw360  |        Caused by: java.lang.ExceptionInInitializerError
sw360  |                at com.liferay.petra.reflect.ReflectionUtil.<clinit>(ReflectionUtil.java:162)
sw360  |                at com.liferay.portal.asm.ASMWrapperUtil.<clinit>(ASMWrapperUtil.java:261)
sw360  |                ... 46 more
sw360  |        Caused by: java.lang.NoSuchFieldException: modifiers
sw360  |                at java.base/java.lang.Class.getDeclaredField(Class.java:2610)
sw360  |                at com.liferay.petra.reflect.ReflectionUtil.<clinit>(ReflectionUtil.java:157)
sw360  |                ... 47 more

```
I attach here the log filtered just for _sw360_ container
[sw360.log](https://github.com/eclipse-sw360/sw360/files/13408058/sw360.log)

**Reply by davidecavestro** [reply]
I still reproduce the same issue locally with default _docker-compose.yml_ and latest docker image ( `ghcr.io/eclipse-sw360/sw360:20231212` as of now)

**Reply by KoukiHama** [reply]
@heliocastro i hope your help.

**Comment by heliocastro** [comment]
I will look closely next week.

**Comment by heliocastro** [comment]
@davidecavestro The docker interface is stable now
I'm closing this discussion
